### PR TITLE
Show exercise participations for course on Team page

### PR DESCRIPTION
**Pull request by madwau**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
It would be nice if the Team page offered some information about a team's activity in the course so that it is easier to keep track of a team's progression for tutors and instructors.

Since teams exist on an exercise-level and not on a course-level, teams with the same short name in the same course are regarded as the "same" team.

_Example of a workflow: A course has five exercises, all of which should be worked on in the same team constellation. Tutors create teams together with the students for the first exercise and the students get started. For subsequent exercises, the instructor uses the "Import teams" feature to clone the existing teams of the first exercise into the second exercise. Cloning will keep the team short names. We know that teams across exercises that have the same short name correspond to the same real-life team._

The Team page will list all of the participations of a team (technically: of all teams with the same short name) for a course. It shows exercise title and release date for context and then details about the participation (start date, number of submissions and status).

### Description
1. Add REST endpoint `GET /courses/:courseId/teams/:teamShortName/with-exercises-and-participations` (get course with all released exercises, for which a team exists by that short name, and their participations for the team)
2. Add component `TeamParticipationTable` that displays the exercises in the course (for which the team exists) and the participation details for the team
3. Include this new component on the Team page and let it be visible for students who belong the the team and for TAs and above (instructors can also look into the submissions by clicking on them, as it is done on the regular participation page; admins see additional id columns useful for debugging)

### Steps for Testing
1. Log in to Artemis as a TA or above
2. Navigate to Course Management
3. Go into a course that has multiple Team Exercises
4. Click on a button "Team" for one of those exercises
5. Verify that the team page shows all exercises for this team short name in the course with participation details

### Screenshots
OnPaste 20200429-170009
**Left:** View as admin (in German), **Right:** View as student (or non-team-owning tutor) (in English)

OnPaste 20200429-171116
**Above:** View as instructor

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/team/team-participation-table/team-participation-table.component.html [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                <ng-template ngx-datatable-header-template>
+                    <span class="datatable-header-cell-wrapper" (click)="controls.onSort('exercise.releaseDate')">
+                        <span class="datatable-header-cell-label bold sortable" jhiTranslate="artemisApp.exercise.releaseDate">
+                            Release Date
+                        </span>
+                        <fa-icon [icon]="controls.iconForSortPropField('exercise.releaseDate')"></fa-icon>
+                    </span>
+                </ng-template>
+                <ng-template ngx-datatable-cell-template let-value="value" class="mr-5">
+                    {{ moment(value) | date: 'dd.MM.yyyy' }}
```
Comment (sleiss) [inline src/main/webapp/app/exercises/shared/team/team-participation-table/team-participation-table.component.html:39]: Apart from this small bug with a undefined date, everything works as expected!

Good work!

Comment (madwau) [reply src/main/webapp/app/exercises/shared/team/team-participation-table/team-participation-table.component.html]: Thanks!

The case of no release date being set is handled now:

OnPaste 20200428-133740

Comment (sleiss) [reply src/main/webapp/app/exercises/shared/team/team-participation-table/team-participation-table.component.html]: Nice!

**Comment by madwau** [comment]
> - Please also show team exercises for the corresponding team without participations (not yet started)
> - Also show the exercises in the team page of the students
> - Hide database IDs for normal users and only show them for admins for debugging purposes

These points are implemented now.

OnPaste 20200429-170009
**Left:** View as admin (in German), **Right:** View as student (or non-team-owning tutor) (in English)

OnPaste 20200429-171116
**Above:** View as instructor

> Tutors and instructors should see the assessment status and the assess / view assessment button (can also be done in a follow up PR)

I will implement this in a follow up PR since this PR has already become quite comprehensive on its own with the suggestions from above.

File: src/main/java/de/tum/in/www1/artemis/service/UserService.java [hunk 2]
```diff
@@ -747,6 +747,17 @@ public User updateUserNotificationReadDate() {
         return userRepository.findAllInGroup(groupName);
     }
 
+    /**
+     * Get all users in a given team
+     *
+     * @param course The course to which the team belongs (acts as a scope for the team short name)
+     * @param teamShortName The short name of the team for which to get all students
+     * @return A list of all users that belong to the team
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/service/UserService.java:755]:
```suggestion
     * @return A set of all users that belong to the team
```

**Comment by madwau** [comment]
> Works fine on the test server 2, I've provided just one minor fix in the documentation.
> 
> Is this bar here intended (and if it is, does it have some purpose I'm missing right now?)?
> image

Thanks for testing!

Yes, the divider bar is intended. It separates the generic exercise information (exercise title and release date) from the participation specific information which only relates to the team one is currently looking at. Also on smaller screens, the two leftmost columns are pinned and only the right part of the table is scrolling (so that the user always knows which exercise the row is for).

**Comment by sleiss** [comment]
> Yes, the divider bar is intended. It separates the generic exercise information (exercise title and release date) from the participation specific information which only relates to the team one is currently looking at. Also on smaller screens, the two leftmost columns are pinned and only the right part of the table is scrolling (so that the user always knows which exercise the row is for).

Thanks for the explanation, makes totally sense, good work!

File: src/main/java/de/tum/in/www1/artemis/web/rest/TeamResource.java [hunk 3]
```diff
@@ -326,4 +331,53 @@ public TeamResource(TeamRepository teamRepository, TeamService teamService, Cour
...
+     *
+     * @param courseId id of the course
+     * @param teamShortName short name of the team (all teams with the short name in the course are seen as the same team)
+     * @return Course with exercises and participations for the team
+     */
+    @GetMapping(value = "/courses/{courseId}/teams/{teamShortName}/with-exercises-and-participations")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Course> getCourseWithExercisesAndParticipationsForTeam(@PathVariable Long courseId, @PathVariable String teamShortName) {
+        log.debug("REST request to get Course {} with exercises and participations for Team with short name {}", courseId, teamShortName);
+        Course course = courseService.findOneWithExercises(courseId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/TeamResource.java:346]: Please use a new method called `courseService.findOneWithTeamExercises` which filters only for team exercises

Comment (madwau) [reply src/main/java/de/tum/in/www1/artemis/web/rest/TeamResource.java:346]: Done

**Comment by madwau** [comment]
> small change request:
> 
> 1. Link to the exercise details page in team page
> 2. Show due date in team page
> 3. Small performance improvement (see comment)

Implemented those three points.

**Pull request merged**